### PR TITLE
DAOS-14834 control: Enable parallel server->engine dRPCs

### PR DESCRIPTION
**Pull request by kjacque**
The idea here is to remove the bottleneck in `daos_server` that serializes dRPC calls, to enable `daos_server` to pass along multiple dRPC calls even if the first one hasn't yet returned. In the current master branch, we have a single dRPC client structure that uses RW locks to control access to its internals. dRPC calls that take a long time can potentially impede other commands.

My proposed solution is to create a new `drpc.ClientConnection` for each command that needs to be sent to the `daos_engine`. Each command is handled on its own connection. We were using a connect->send->disconnect pattern on the client connection anyway.

Required-githooks: true

### Before requesting gatekeeper:

* [x] Two review approvals and any prior change requests have been resolved.
* [x] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [x] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [x] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [x] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/control/server/instance_drpc.go [hunk 1]
```diff
@@ -29,28 +29,31 @@ var (
 	errEngineNotReady = errors.New("engine not ready yet")
 )
 
-func (ei *EngineInstance) setDrpcClient(c drpc.DomainSocketClient) {
+func (ei *EngineInstance) setDrpcSocket(sock string) {
 	ei.Lock()
 	defer ei.Unlock()
-	ei._drpcClient = c
+	ei._drpcSocket = sock
```
Comment (tanabarr) [inline src/control/server/instance_drpc.go:35]: nit: don't see a need for the underscore variable prefix

**Review by tanabarr — DISMISSED** [review]
LGTM, any issues in practice?

File: src/control/server/instance_drpc.go (hunk 1 above)
Comment (kjacque) [reply src/control/server/instance_drpc.go:35]: Based on a comment on the struct definition, the reason this was initially done was to mark variables that need to be behind a lock. As this variable subject to the same limitations _drpcClient originally was, I kept the underscore.

File: src/control/server/instance_drpc.go [hunk 2]
```diff
@@ -29,28 +29,31 @@ var (
...
 	ei.RLock()
 	defer ei.RUnlock()
-	if ei._drpcClient == nil {
-		return nil, errDRPCNotReady
+	return ei._drpcSocket
+}
+
+func (ei *EngineInstance) getDrpcClient() drpc.DomainSocketClient {
+	if ei.getDrpcClientFn == nil {
+		ei.getDrpcClientFn = drpc.NewClientConnection
```
Comment (mjmac) [inline src/control/server/instance_drpc.go:46]: Isn't this racy? Could be called from multiple goroutines, I think...

File: src/control/server/instance_drpc.go [hunk 3]
```diff
@@ -29,28 +29,31 @@ var (
...
-	ei._drpcClient = c
+	ei._drpcSocket = sock
 }
 
-func (ei *EngineInstance) getDrpcClient() (drpc.DomainSocketClient, error) {
+func (ei *EngineInstance) getDrpcSocket() string {
 	ei.RLock()
 	defer ei.RUnlock()
-	if ei._drpcClient == nil {
-		return nil, errDRPCNotReady
```
Comment (mjmac) [inline src/control/server/instance_drpc.go:42]: If you're getting rid of this logic, how will we detect/handle requests that are made before the engine has sent its ready upcall? I agree that now is a good time to rework some of this fiddly logic, though.

**Review by mjmac — COMMENTED** [review]
Interesting to see how this was achieved with relatively minor changes to the logic. I do think that most of the really tricky concerns with concurrent dRPCs have been alleviated by locking at the pool level and serialization of handlers (e.g. PoolCreate). Might be a good opportunity to go even further to streamline and simplify some of this instance management code.

File: src/control/server/instance_drpc.go (hunk 2 above)
Comment (kjacque) [reply src/control/server/instance_drpc.go:46]: Good catch, I'll lock around this.

Comment (kjacque) [reply src/control/server/instance_drpc.go:48]: Fixed in the latest push.

File: src/control/server/instance_drpc.go (hunk 3 above)
Comment (kjacque) [reply src/control/server/instance_drpc.go:42]: Fixed in the latest push. I moved this check earlier in the CallDrpc flow.

**Review by tanabarr — APPROVED** [review]

**Comment by kjacque** [comment]
Test failures are unrelated.
- DaosCoreTest failure on UCX: https://daosio.atlassian.net/issues/DAOS-15275
- DmgStorageQueryTest failure: https://daosio.atlassian.net/browse/DAOS-15808
- The other one is a DaosCoreTest rebuild timeout, for which we have multiple tickets.

**Comment by tanabarr** [comment]
have we verified that this fixes the customer issue or is that only possible after landing?

**Review by liw — APPROVED** [review]
Thanks, @kjacque. AFAICS, it looks good to me.

**Comment by kjacque** [comment]
> have we verified that this fixes the customer issue or is that only possible after landing?

Hasn't been verified (I don't think we have a reliable reproducer for the issue). However at least part of the problem was the server being blocked from further dRPCs to the engine by one dRPC hogging the channel, so that should be resolved by this PR.

**Comment by kjacque** [comment]
@daos-stack/daos-gatekeeper Test failures are documented above. Can this be force-landed please?

**Review by mjmac — APPROVED** [review]
LGTM. Nice work, Kris. The unit test looks really solid, too.

**Pull request merged**